File: 1-Easy/Panel count/panel-count.java
import java.util.Scanner;

class Property {
    private String name;
    private String value;

    public Property(String name, String value) {
        this.name = name;
        this.value = value;
    }

    public String getName() {
        return name;
    }

    public String getValue() {
        return value;
    }
}

class Person {
    private String name;
    private Property[] properties;

    public Person(String line, String[] propertyNames) {
        String[] elements = line.split(" ");
        this.name = elements[0];
        this.properties = new Property[propertyNames.length];

        for (int i = 0; i < propertyNames.length; i++) {
            properties[i] = new Property(propertyNames[i], elements[i + 1]);
        }
    }

    public String getName() {
        return name;
    }

    public Property[] getProperties() {
        return properties;
    }
}

class Formula {
    private Property[] properties;

    public Formula(String line) {

        // Split the line into properties
        String[] elements = line.split(" AND ");
        this.properties = new Property[elements.length];

        for (int i = 0; i < elements.length; i++) {
            // Split the property into name and value
            String[] property = elements[i].split("=");

            properties[i] = new Property(property[0], property[1]);
        }
    }

    public Property[] getProperties() {
        return properties;
    }
}

public class Solution {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        String[] propertyNames = readPropertyNames(in);
        Person[] people = readPeople(in, propertyNames);
        Formula[] formulas = readFormulas(in);

        printResult(formulas, people);

        in.close();
    }

    private static String[] readPropertyNames(Scanner in) {
        int propertyCount = in.nextInt();
        String[] propertyNames = new String[propertyCount];

        if (in.hasNextLine()) {
            in.nextLine();
        }

        for (int i = 0; i < propertyCount; i++) {
            propertyNames[i] = in.nextLine();
        }

        return propertyNames;
    }

    private static Person[] readPeople(Scanner in, String[] propertyNames) {
        int personCount = in.nextInt();
        Person[] people = new Person[personCount];

        if (in.hasNextLine()) {
            in.nextLine();
        }

        for (int i = 0; i < personCount; i++) {
            people[i] = new Person(in.nextLine(), propertyNames);
        }

        return people;
    }

    private static Formula[] readFormulas(Scanner in) {
        int formulaCount = in.nextInt();
        Formula[] formulas = new Formula[formulaCount];

        if (in.hasNextLine()) {
            in.nextLine();
        }
        
        for (int i = 0; i < formulaCount; i++) {
            formulas[i] = new Formula(in.nextLine());
        }

        return formulas;
    }

    // Iterates through the formulas and people, incrementing matchingCount if the given person's properties fit the formula
    // Finally, it logs the number of people who match to the console per formula
    private static void printResult(Formula[] formulas, Person[] people) {
        for (Formula formula : formulas) {
            int matchingCount = 0;

            for (Person person : people) {
                if (matchesFormula(formula, person)) {
                    matchingCount++;
                }
            }
            
            System.out.println(matchingCount);
        }
    }

    // Iterates through the formula's and person's properties, and checks if the names and values match. 
    private static boolean matchesFormula(Formula formula, Person person) {
        Property[] formulaProperties = formula.getProperties();
        Property[] personProperties = person.getProperties();

        for (Property formulaProperty : formulaProperties) {
            boolean propertyMatched = false;

            for (Property personProperty : personProperties) {
                boolean nameMatches = formulaProperty.getName().equals(personProperty.getName());
                boolean valueMatches = formulaProperty.getValue().equals(personProperty.getValue());

                if (nameMatches && valueMatches) {
                    propertyMatched = true;
                    break;
                }
            }

            if (!propertyMatched) {
                return false;
            }
        }
        
        return true;
    }
}

File: 1-Easy/Panel count/panel-count.ts
class Property {
    private name: string;
    private value: string;

    public constructor(name: string, value: string) {
        this.name = name;
        this.value = value;
    }

    public getName(): string {
        return this.name;
    }

    public getValue(): string {
        return this.value;
    }
}

class Person {
    private name: string;
    private properties: Property[];

    public constructor(line: string, propertyNames: string[]) {
        const elements = line.split(" ");
        this.name = elements[0];
        this.properties = [];

        for (let i = 0; i < propertyNames.length; i++) {
            this.properties[i] = new Property(propertyNames[i], elements[i + 1]);
        }
    }

    public getName(): string {
        return this.name;
    }

    public getProperties(): Property[] {
        return this.properties;
    }
}

class Formula {
    private properties: Property[];

    public constructor(line: string) {
        // Split the line into properties
        const elements = line.split(" AND ");
        this.properties = [];

        for (let i = 0; i < elements.length; i++) {
            // Split the property into name and value
            const property = elements[i].split("=");
            this.properties[i] = new Property(property[0], property[1]);
        }
    }

    public getProperties(): Property[] {
        return this.properties;
    }
}

class Solution {
    public static main() {
        const propertyNames = this.readPropertyNames();
        const people = this.readPeople(propertyNames);
        const formulas = this.readFormulas();

        this.printResult(formulas, people);
    }

    private static readPropertyNames(): string[] {
        const propertyCount = parseInt(readline());
        const propertyNames = [];

        for (let i = 0; i < propertyCount; i++) {
            propertyNames[i] = readline();
        }

        return propertyNames;
    }

    private static readPeople(propertyNames: string[]): Person[] {
        const personCount = parseInt(readline());
        const people = [];

        for (let i = 0; i < personCount; i++) {
            people[i] = new Person(readline(), propertyNames);
        }

        return people;
    }

    private static readFormulas(): Formula[] {
        const formulaCount = parseInt(readline());
        const formulas = [];

        for (let i = 0; i < formulaCount; i++) {
            formulas[i] = new Formula(readline());
        }

        return formulas;
    }

    // Iterates through the formulas and people, incrementing matchingCount if the given person's properties fit the formula
    // Finally, it logs the number of people who match to the console per formula
    private static printResult(formulas: Formula[], people: Person[]) {
        for (const formula of formulas) {
            let matchingCount: number = 0;

            for (const person of people) {
                if (this.matchesFormula(formula, person)) {
                    matchingCount++;
                }
            }
            
            console.log(matchingCount);
        }
    }

    // Iterates through the formula's and person's properties, and checks if the names and values match. 
    private static matchesFormula(formula: Formula, person: Person): boolean {
        const formulaProperties = formula.getProperties();
        const personProperties = person.getProperties();

        for (const formulaProperty of formulaProperties) {
            let propertyMatched: boolean = false;

            for (const personProperty of personProperties) {
                const nameMatches = formulaProperty.getName() === personProperty.getName();
                const valueMatches = formulaProperty.getValue() === personProperty.getValue();

                if (nameMatches && valueMatches) {
                    propertyMatched = true;
                    break;
                }
            }

            if (!propertyMatched) {
                return false;
            }
        }
        
        return true;
    }
}

Solution.main();